lib/madwf_ml.cpp

```diff
@@ -0,0 +1,348 @@
+
+      alpha = 0;
+      double root_min = poly4(a, 0);
+      for (auto r : rs) {
+        double eval = poly4(a, r);
+        if (root_min > eval) {
+          root_min = eval;
+          alpha = r;
+        }
+      }
+
+      axpby(device_param, 0.0f, device_param, -alpha, P);
+      if (tune_suppressor) { mu -= alpha * pmu; }
+
+      if (getVerbosity() >= QUDA_SUMMARIZE) {
+        printfQuda("grad min iter %05d: %04d chi2 = %8.4e, chi2 %% = %8.4e, alpha = %+8.4e, mu = %+8.4e\n", comm_rank(),
+                   iteration, chi2, chi2 / residual, alpha, mu);
+      }
+    }
+
+    trained = true;
+
+    if (getVerbosity() >= QUDA_VERBOSE) { printfQuda("Training finished ...\n"); }
+    count = 0;
+    for (auto phi: B) {
+      double ind_chi2 = cost(ref, base, chi, *phi);
+      double phi2 = blas::norm2(*phi);
+      if (getVerbosity() >= QUDA_VERBOSE) {
+        printfQuda("chi2 %03d %% = %8.4e, phi2 = %8.4e\n", count, ind_chi2 / phi2, phi2);
+      }
+      count++;
+    }
+
+    for (auto pb: B) { delete pb; }
+
+    // Broadcast the trained parameters
+    host_param = device_param.to_host();
+    comm_broadcast(host_param.data(), host_param.size() * sizeof(transfer_float));
+    device_param.from_host(host_param);
+
+    commGlobalReductionPop();
+    profile.TPSTOP(QUDA_PROFILE_TRAINING);
+
+    if (param.madwf_param_save) {
+      profile.TPSTART(QUDA_PROFILE_IO);
+      if (comm_rank() == 0) { save_parameter(Ls, Ls_base); } // Only rank zero write out to the disk
+      comm_barrier();
+      profile.TPSTOP(QUDA_PROFILE_IO);
+    }
+
+  }
+
+  void MadwfAcc::save_parameter(int Ls, int Ls_base) {
+    std::vector<transfer_float> host_param = device_param.to_host();
+
+    std::string save_param_path(param.madwf_param_outfile);
+    char cstring[512];
+    sprintf(cstring, "/madwf_trained_param_ls_%02d_%02d_mu_%.3f.dat", Ls, Ls_base, mu);
+    save_param_path += std::string(cstring);
+    FILE *fp = fopen(save_param_path.c_str(), "w");
+    if (!fp) { errorQuda("Unable to open file %s\n", save_param_path.c_str()); }
+    size_t fwrite_count = fwrite(host_param.data(), sizeof(transfer_float), host_param.size(), fp);
+    fclose(fp);
+    if (fwrite_count != host_param.size()) {
+      errorQuda("Unable to write trained parameters to %s (%lu neq %lu).\n", save_param_path.c_str(), fwrite_count,
+          host_param.size());
+    }
+    if (getVerbosity() >= QUDA_VERBOSE) {
+      printfQuda("Trained parameters saved to %s ...\n", save_param_path.c_str());
+    }
+  }
+
+  void MadwfAcc::load_parameter(int Ls, int Ls_base) {
+    constexpr int complex_matrix_size = static_cast<int>(transfer_t); // spin by spin
+    size_t param_size = Ls * Ls_base * complex_matrix_size * 2;
+    std::vector<transfer_float> host_param(param_size);
+
+    char param_file_name[512];
+    // Note that all ranks load from the same file.
+    sprintf(param_file_name, "/madwf_trained_param_ls_%02d_%02d_mu_%.3f.dat", Ls, Ls_base, mu);
```

Comment:
See comment above about each rank reading from the same file (admittedly, this is less dangerous than the write)

Reply:
And yes here each rank will read from the same file.